lib/matplotlib/backend_bases.py

```diff
@@ -3596,29 +3593,62 @@ def _message_cbk(self, event):
         """Captures the 'tool_message_event' to set the message on the toolbar"""
         self.set_message(event.message)
 
-    def _tool_triggered_cbk(self, event):
+    def _tool_toggled_cbk(self, event):
         """Captures the 'tool-trigger-toolname
 
         This only gets used for toggled tools
         """
-        if event.sender is self:
-            return
+        self.toggle_toolitem(event.tool.name, event.tool.toggled)
 
-        self.toggle_toolitem(event.tool.name)
-
-    def _add_tool_cbk(self, event):
-        """Captures 'tool_added_event' and adds the tool to the toolbar"""
-        image = self._get_image_filename(event.tool.image)
-        toggle = getattr(event.tool, 'toggled', None) is not None
-        self.add_toolitem(event.tool.name,
-                          event.data['group'],
-                          event.data['position'],
-                          image,
-                          event.tool.description,
-                          toggle)
+    def add_tools(self, tools):
+        """ Add multiple tools to `Navigation`
+
+        Parameters
+        ----------
+        tools : List
+            List in the form
+            [[group1, [name1, name2 ...]][group2...]]
+            where group1 is the name of the group where the
+            Tool1, Tool2... are going to be added, and name1, name2... are the
+            names of the tools
+        """
+
+        for group, grouptools in tools:
+            for position, tool in enumerate(grouptools):
+                self.add_tool(tool, group, position)
+
+    def add_tool(self, tool, group, position=-1, name=None, **kwargs):
+        """Adds a tool to the toolbar
+
+        Parameters
+        ----------
+        tool : string, tool
+            The name or the type of tool to add.
+        group : string
+            The name of the group to add this tool to.
+        position : int
+            the relative position within the group to place this tool.
+        name : string (optional)
+            If given, and the above fails, we use this to create a new tool of
+            type given by tool, and use this as the name of the tool.
+        """
+        t = self.navigation.get_tool(tool, False)
+        if t is None:
+            if not isinstance(tool, type):
```

Comment:
This is not right!. if you try passing a tool class directly to the toolbar, this will raise
```
/home/fariza/workspace/matplotlib/lib/matplotlib/backend_bases.py:3570: UserWarning: <class '__main__.ListTools'> is not a tool controlled by Navigation
  warnings.warn("%s is not a tool controlled by Navigation" % name)
Traceback (most recent call last):
  File "/home/fariza/workspace/matplotlib/examples/user_interfaces/navigation.py", line 59, in <module>
    fig.canvas.manager.toolbar.add_tool(ListTools, 'foo', name='dude')
  File "/home/fariza/workspace/matplotlib/lib/matplotlib/backend_bases.py", line 3640, in add_tool
    t = self.navigation.add_tool(name, tool)
  File "/home/fariza/workspace/matplotlib/lib/matplotlib/backend_bases.py", line 3430, in add_tool
    self._tools[name] = tool_cls(self, name)
TypeError: type() takes 1 or 3 arguments
```